### PR TITLE
Username Password + Namespaces + Timeout

### PR DESCRIPTION
**Pull request by dulvac**
fixes #27 
fixes #28

* Implemented feedback on the other two PRs
* Added a timeout global config, backed by `com.bettercloud.vault` java driver retry mechanism
* Updated project dependencies
* Refactoring of clunky `VaultAccessor` init to use `VaultConfig` directly.

### CODE REVIEW

**Comment by dulvac** [comment]
@casz , Thanks a lot for looking at these. We've added a few more improvements and it's getting hard to keep track of branches. Would you be willing to accept these improvements as one change? that would simplify things a lot.

The big one, from a value point of view is the `VaultUsernamePasswordCredential` addition

**Comment by dulvac** [comment]
cc @volteanu

File: pom.xml [hunk 1]
```diff
@@ -9,7 +9,7 @@
   </parent>
   <groupId>com.datapipe.jenkins.plugins</groupId>
   <artifactId>hashicorp-vault-plugin</artifactId>
-  <version>2.2.1-SNAPSHOT</version>
+  <version>3.0.0-SNAPSHOT</version>
```
Comment (casz) [inline pom.xml]: Please do not fiddle with the version number 😅

**Comment by casz** [comment]
Really sucks you created a new PR 😂 but its fine

File: src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java [hunk 2]
```diff
@@ -127,9 +129,26 @@ public void setVaultAccessor(VaultAccessor vaultAccessor) {
...
         }
 
+        VaultConfig vaultConfig = null;
+        try {
+            vaultConfig = new VaultConfig()
+                    .address(configuration.getVaultUrl())
+                    .sslConfig(new SslConfig().verify(configuration.isSkipSslVerification()).build());
+
+            if (configuration.getVaultNamespace() != null && !configuration.getVaultNamespace().isEmpty()) {
+                    vaultConfig.nameSpace(configuration.getVaultNamespace());
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java]: 🤕 why 8 whitespace?

File: src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java [hunk 3]
```diff
@@ -58,23 +66,31 @@ public VaultConfiguration mergeWithParent(VaultConfiguration parent) {
...
+            result.setVaultNamespace(parent.getVaultNamespace());
+        }
+        if (null == result.timeout) {
+            result.setTimeout(parent.getTimeout());
+        }
         result.failIfNotFound = failIfNotFound;
         return result;
     }
 
+
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java]: 👻 white space?

File: src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java [hunk 4]
```diff
@@ -58,23 +66,31 @@ public VaultConfiguration mergeWithParent(VaultConfiguration parent) {
...
+
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java]: 👻 white space?

File: src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java [hunk 5]
```diff
@@ -51,7 +52,7 @@
 import jenkins.model.GlobalConfiguration;
 
 public class VaultConfigurationIT {
-   @Rule
+    @Rule
```
Comment (casz) [inline src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java]: 👻 white space?

File: src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java [hunk 6]
```diff
@@ -60,26 +61,34 @@
...
 
-      SystemCredentialsProvider.getInstance().setDomainCredentialsMap(Collections.singletonMap(Domain.global(), Arrays
-            .asList(GLOBAL_CREDENTIAL_1, GLOBAL_CREDENTIAL_2)));
 
-      this.project = jenkins.createFreeStyleProject("test");
-   }
+        GLOBAL_CREDENTIAL_1 = createTokenCredential(GLOBAL_CREDENTIALS_ID_1);
+        GLOBAL_CREDENTIAL_2 = createTokenCredential(GLOBAL_CREDENTIALS_ID_2);
+
+        SystemCredentialsProvider.getInstance().setDomainCredentialsMap(Collections.singletonMap(Domain.global(), Arrays
```
Comment (casz) [inline src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java]:
```suggestion
        SystemCredentialsProvider.getInstance().setDomainCredentialsMap(Collections.singletonMap(Domain.global(),
```

File: src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java [hunk 7]
```diff
@@ -60,26 +61,34 @@
...
+                .asList(GLOBAL_CREDENTIAL_1, GLOBAL_CREDENTIAL_2)));
```
Comment (casz) [inline src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java]:
```suggestion
                Arrays.asList(GLOBAL_CREDENTIAL_1, GLOBAL_CREDENTIAL_2)));
```

**Comment by casz** [comment]
I think once all PRs are merged/closed I will set up some default formatting 😓

File: src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultUsernamePasswordCredentialImpl.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        return Secret.fromString(getValue(this.passwordKey));
+    }
+
+
+    private String getValue(String valueKey) {
+        return getVaultSecret(this.getPath(), valueKey, this.getEngineVersion());
+    }
+
+    private static String getVaultSecret(String secretPath, String secretKey, Integer engineVersion) {
+        if (Jenkins.getInstanceOrNull() == null) {
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/credentials/common/VaultUsernamePasswordCredentialImpl.java]: this cannot and should never happen

File: src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java [hunk 9]
```diff
@@ -127,9 +129,26 @@ public void setVaultAccessor(VaultAccessor vaultAccessor) {
...
+            }
+        } catch (VaultException e) {
+            throw new VaultPluginException("Could not set up VaultConfig.", e);
+        }
         VaultCredential credential = retrieveVaultCredentials(build);
 
-        vaultAccessor.init(url, credential, configuration.isSkipSslVerification());
+        vaultAccessor.setConfig(vaultConfig);
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java]: This could be a builder pattern

**Comment by casz** [comment]
Sorry for the headache @dulvac but would you be willing to fix up conflicts and look into the requested changes?

File: src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java (hunk 9 above)
Comment (casz) [reply src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java]: Sorry disregard the above comment, I did not see the full context when looking at the code 😆

**Comment by dulvac** [comment]
@casz, sorry for radio silence. I can't tackle this at the moment and the conflicts go quite deep as I've seem to have lost some commit races :)

I'll get back to this is a few days, hopefully, if that's ok with you.

**Comment by casz** [comment]
:sob: to not allowing maintainer commit access.

**Pull request closed without merging**